### PR TITLE
random-util: remove RDRAND usage

### PR DESCRIPTION
**Pull request by zx2c4**
```
    /dev/urandom is seeded with RDRAND. Calling genuine_random_bytes(...,
    ..., 0) will use /dev/urandom as a last resort. Hence, we gain nothing
    here by having our own RDRAND wrapper, because /dev/urandom already is
    based on RDRAND output, even before /dev/urandom has fully initialized.
    
    Furthermore, RDRAND is not actually fast! And on each successive
    generation of new x86 CPUs, from both AMD and Intel, it just gets
    slower.
    
    This commit simplifies things by just using /dev/urandom in cases where
    we before might use RDRAND, since /dev/urandom will always have RDRAND
    mixed in as part of it.
    
    And above where I say "/dev/urandom", what I actually mean is
    GRND_INSECURE, which is the same thing but won't generate warnings in
    dmesg.
```

### CODE REVIEW

**Comment by zx2c4** [comment]
This is marked as a draft while https://github.com/systemd/systemd/pull/22667 is reviewed first and separately.

**Comment by poettering** [comment]
We so far are using RDRAND for generating UUIDs, i.e. where we want high entropy, but not quite cryptographic quality, and where the kernel random pool is not seeded with RDRAND, because distros traditionally didn't do that, specifically Debian (they set CONFIG_RANDOM_TRUST_CPU=no for the longest time, though that might have changed now). We need a lot of UUIDs (we pass one to each service invocation, as an identifier of a specific instance of a service), hence having to wait for being able to generate UUIDs at boot really sucks.

Is CONFIG_RANDOM_TRUST_CPU=no still a thing? Is this option removed from the kernel now, too?

Usually, we don't hard rely on kernel features for a few years. Are we really there yet with this?

**Comment by bluca** [comment]
>     /dev/urandom is seeded with RDRAND. Calling genuine_random_bytes(...,
    ..., 0) will use /dev/urandom as a last resort. Hence, we gain nothing
    here by having our own RDRAND wrapper, because /dev/urandom already is
    based on RDRAND output, even before /dev/urandom has fully initialized

Is this true on all kernels or just new versions? As per readme we need to support from 3.12 and up

**Comment by poettering** [comment]
> Is this true on all kernels or just new versions? As per readme we need to support from 3.13 and up

Which I'd be fine to raise btw, 3.13 is 2014, i.e. 8 years ago... I think supporting 5y old kernels is already more than enough.

**Comment by zx2c4** [comment]
> where the kernel random pool is not seeded with RDRAND, because distros traditionally didn't do that, specifically Debian (they set CONFIG_RANDOM_TRUST_CPU=no for the longest time, though that might have changed now). 

This is actually a common misconception. CONFIG_RANDOM_TRUST_CPU does not control whether RDSEED is mixed into the pool. RDSEED is always mixed into the pool. Rather, it just controls whether it's credited for it -- whether the RNG unblocks and becomes fully operational on the basis of RDRAND. So even if you set CONFIG_RANDOM_TRUST_CPU=no, it'll still use RDRAND for GRND_INSECURE and such.

(I was hoping we'd get https://github.com/systemd/systemd/pull/22667 in first and then you'd review this after but I guess that didn't happen, so I'll mark this as non-draft now.)

**Comment by poettering** [comment]
> This is actually a common misconception. CONFIG_RANDOM_TRUST_CPU does not control whether RDSEED is mixed into the pool. RDSEED is always mixed into the pool.

Ah, interesting, and that was the case already in 5y old kernels?

So this would only leave one benefit of calling RDSEED then: that we don't see that annoying /dev/urandom kernel log message on old kernels, right? Upcoming kernels don't generate that message at all anymore, right? And all kernels that support GRND_INSECURE don't either (which was added in ~2019 or so), because we go via getrandom() there?

So the question would be whether it's worth keeping the RDSEED support in place to suppress the ugly log message on kernels from before GRND_INSECURE? That's really all it boils down to, right?

**Comment by poettering** [comment]
> Right, exactly. Do you want to maintain complex and historically error-prone code just for the sake or suppressing a log message on ancient kernels that likely won't be used with newer systemds in most cases? For me the choice is an easy one.

I added the RDSEED stuff because people filed all kinds of bugs against systemd here and in distro bug trackers because of that kernel log message under the assumption systemd was doing something wrong there and was crap. GRND_INSECURE was added to kernel 5.6, i.e. April 2020, which is not even 2y ago, so if new systemd would be used with kernels before that people will see the warning. As I know the Debian people they care about really old kernels and about the cosmetics of their kernel logs, so I figure they at least will not like this change.

**Comment by bluca** [comment]
> > Right, exactly. Do you want to maintain complex and historically error-prone code just for the sake or suppressing a log message on ancient kernels that likely won't be used with newer systemds in most cases? For me the choice is an easy one.
> 
> I added the RDSEED stuff because people filed all kinds of bugs against systemd here and in distro bug trackers because of that kernel log message under the assumption systemd was doing something wrong there and was crap. GRND_INSECURE was added to kernel 5.6, i.e. April 2020, which is not even 2y ago, so if new systemd would be used with kernels before that people will see the warning. As I know the Debian people they care about really old kernels and about the cosmetics of their kernel logs, so I figure they at least will not like this change.

It's especially the embedded world, as they tend to get stuck on ancient kernel versions, not their fault but due to bad hardware vendors - see all that noise we got due to the /proc hidepid mount option causing a warn-once on old kernels last year or so...

**Comment by zx2c4** [comment]
Can't they just disable that warning if they need? Debian is already stuck on old systemd + old kernel. When they update systemd, they'll have a new kernel too. What x86 distros have old kernel + new systemd? All I can think of are situations with arm vendor kernels, and there's no rdrand there. It really seems like this is worth a shot, and that practically speaking, impact should be pretty minimal. And if people complain we can just revert.

**Comment by zx2c4** [comment]
> It's especially the embedded world

There is no rdrand code path for those. Non-issue.

**Comment by zx2c4** [comment]
This is now rebased on https://github.com/systemd/systemd/pull/22667, which should show a bit better the actual case where we wind up using RDRAND, and thus more precisely shows what we're getting rid of here.

As mentioned in my last message, I don't think we're getting rid of anything important in practice with this patch.

**Comment by zx2c4** [comment]
Looking at the CI, it days CentOS 8S failed, but I can't see how that failure is related to this commit. I think it might be a "flake"? Or something else and maybe I should rebase on master to get the green checkbox?

Either way, I think this PR is good to go and ready to review/merge/whatever.

**Comment by bluca** [comment]
> /cc @mbiebl

I don't think this is an issue for any Debian usage. There are no Debian x86 releases with ancient kernels but new systemd. At most you can have new systemd via backports with the kernel shipped in stable, but that's v5.10 already, so we are good. You'd need to have a user manually keeping oldstable x86 kernel on stable userspace, plus systemd from backports, but that seems excessively unlikely and it would be an unsupported setup as far as we are concerned.

These issues tend to come up in the embedded world, ie openembedded/yocto and similar. I imagine you were thinking about the procfs hidepid stuff, and that was from yocto/buildroot/etc. But as Jason noted this only affects x86, so although there are x86 builds for those, it would be a weird setup even in that context. It might happen of course, and perhaps 2 years is a bit on the short side of the scale.

Overall it seems to me this is something that we could risk getting in v252 (ie, second half of the year), with little risk of backlash. We can always revert and further postpone it if it happens.

**Comment by zx2c4** [comment]
> These issues tend to come up in the embedded world, ie openembedded/yocto and similar. I imagine you were thinking about the procfs hidepid stuff, and that was from yocto/buildroot/etc. But as Jason noted this only affects x86, so although there are x86 builds for those, it would be a weird setup even in that context. It might happen of course, and perhaps 2 years is a bit on the short side of the scale.

It would be really weird, even in yocto x86 world. But something to keep in mind for the embedded use cases: those extreme edge cases doing weird things are in the best possible position for simply _turning off the kmsgs_ with the kernel command line, precisely _because_ they're embedded systems doing weird things. In other words, old x86 systems that are trying to forcibly upgrade to new systemd can just adjust the thing that RDRAND was there to compensate for. So that seems like another non-issue.

Anyway, just my 2¢; feel free to merge this whenever makes most sense to you. It's obviously not urgent, though it is a maintenance burden of sorts and who knows when you'll get bit by the next irksome AMD RDRAND bug or whatever. I've just rebased this on main, per the label.

**Comment by poettering** [comment]
I am inclined to merge this then, unless anyone is loudly opposed?

**Comment by poettering** [comment]
hmm, docs/RANDOM_SEEDS.md needs updating I guess. It talks a lot about RDSEED still...

File: src/basic/random-util.c
Comment (medhefgo) [inline src/basic/random-util.c:104]: While this code is touched, I'd love to see this declaration be moved into the other `#if`.

Comment (zx2c4) [reply src/basic/random-util.c:104]: There's a statement in between the two, and afaict, systemd declares variables before statements.

**Comment by zx2c4** [comment]
> hmm, docs/RANDOM_SEEDS.md needs updating I guess. It talks a lot about RDSEED still...

Fixed.

File: src/basic/random-util.c
Comment (medhefgo) [reply src/basic/random-util.c:104]: That restriction was removed some time ago.

Comment (zx2c4) [reply src/basic/random-util.c:104]: Oh. Maybe send another PR for that? I don't see how it's related to this, and it'd be nice to have this stand-alone revertable in case things don't go as planned.

File: docs/RANDOM_SEEDS.md
Comment (bluca) [inline docs/RANDOM_SEEDS.md]: Is this an actual thing? GRND_INSECURE being unavailable because of configuration/whatnot by the hypervisor provider?

Comment (zx2c4) [reply docs/RANDOM_SEEDS.md]: Not quite. This may have just been a weird edit and maybe I should remove the bullet point all together? I was thinking that `virtio-rng` might initialize the rng pretty fast, so you don't even need GRND_INSECURE, while GRND_INSECURE fills the needs of RDRAND.

Do you want me to just remove the bullet point?

Comment (zx2c4) [reply docs/RANDOM_SEEDS.md]: Actually, the original text is fine. I'll restore that chunk. RDRAND _is_ relevant here because it can seed the kernel RNG.

Comment (bluca) [reply docs/RANDOM_SEEDS.md]: What about:

```
2. Virtualized environments that lack virtio-rng and a kernel without `GRND_INSECURE`,
   outside of test environments. Tough luck. Talk to your hosting provider, and ask them to
   fix this, or switch your VM to a newer kernel that provides `GRND_INSECURE` (v5.6
   and newer).
```

Comment (bluca) [reply docs/RANDOM_SEEDS.md]: sure, that works too

Comment (zx2c4) [reply docs/RANDOM_SEEDS.md]: That's what I was thinking at first too, but I just reread the whole section, and it's actually a discussion of bootstrapping the RNG to be secure, so I think the mention of GRND_INSECURE doesn't fit here, whilst the mention of RDRAND does, because:

```c
	if (arch_init && trust_cpu && !crng_ready()) {
		crng_init = 2;
		pr_notice("crng init done (trusting CPU's manufacturer)\n");
	}
```

**Comment by zx2c4** [comment]
CI is all green now. Any thing else holding this up that I should address?

**Comment by zx2c4** [comment]
Added a `NEWS` entry about this to mention the kmsg change.

File: NEWS
Comment (poettering) [inline NEWS]: indentation is off

Comment (zx2c4) [reply NEWS]: Tabs vs spaces again... Fixing.

**Comment by zx2c4** [comment]
Fixed `\t` vs ` ` and rebased. Should be good to go now.

**Comment by poettering** [comment]
excellent! thanks!

**Review by bluca — APPROVED** [review]

**Pull request merged**
